// exupery-resources/dist/queries/unguaranteed/execute_query_executable.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.$$ = void 0;
const _easync = __importStar(require("exupery-core-async"));
const _ei = __importStar(require("exupery-core-internals"));
const node_child_process_1 = require("node:child_process");
// ---------------------------------------------------------------
/**
 *
 * The executable being executed is assumed to be side effect free
 * There is no way to give guarantees about that though
 */
const $$ = ($p) => {
    const args = $p.args.__get_raw_copy();
    return _easync.__create_unguaranteed_query({
        'execute': (on_value, on_exception) => {
            const child = (0, node_child_process_1.spawn)($p.program, args, {
                shell: false, // ✅ no implicit parsing
            });
            let stdoutData = "";
            let stderrData = "";
            child.stdout.on("data", chunk => {
                stdoutData += chunk.toString("utf8");
            });
            child.stderr.on("data", chunk => {
                stderrData += chunk.toString("utf8");
            });
            child.on("error", err => {
                on_exception(_ei.block(() => {
                    return ['failed to spawn', {
                            message: err instanceof Error ? err.message : `${err}`
                        }];
                }));
            });
            child.on("close", exitCode => {
                if (exitCode === 0) {
                    on_value({
                        stdout: stdoutData,
                    });
                }
                else {
                    on_exception(_ei.block(() => {
                        return ['non zero exit code', {
                                exitCode: exitCode ?? -1,
                                stderr: stderrData,
                            }];
                    }));
                }
            });
        }
    });
};
exports.$$ = $$;
