// pareto-fountain-pen/dist/generated/implementation/schemas/block/marshall.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Node = exports.Group_Part = exports.Group = exports.Directory = exports.Block_Part = exports.Block = void 0;
const _pa = __importStar(require("exupery-core-alg"));
const Block = ($, $p) => ['list', $.map(($) => (0, exports.Block_Part)($, {
        'value serializers': $p['value serializers'],
    }))];
exports.Block = Block;
const Block_Part = ($, $p) => ['state', _pa.cc($, ($) => {
        switch ($[0]) {
            case 'indent': return _pa.ss($, ($) => ({
                'state': "indent",
                'value': (0, exports.Group)($, {
                    'value serializers': $p['value serializers'],
                }),
            }));
            case 'nothing': return _pa.ss($, ($) => ({
                'state': "nothing",
                'value': ['nothing', null],
            }));
            case 'optional': return _pa.ss($, ($) => ({
                'state': "optional",
                'value': ['optional', $.transform(($) => ['set', (0, exports.Block_Part)($, {
                            'value serializers': $p['value serializers'],
                        })], () => ['not set', null])],
            }));
            case 'snippet': return _pa.ss($, ($) => ({
                'state': "snippet",
                'value': ['text', ({
                        'delimiter': ['quote', null],
                        'value': $,
                    })],
            }));
            case 'sub block': return _pa.ss($, ($) => ({
                'state': "sub block",
                'value': (0, exports.Block)($, {
                    'value serializers': $p['value serializers'],
                }),
            }));
            default: return _pa.au($[0]);
        }
    })];
exports.Block_Part = Block_Part;
const Directory = ($, $p) => ['dictionary', $.map(($) => (0, exports.Node)($, {
        'value serializers': $p['value serializers'],
    }))];
exports.Directory = Directory;
const Group = ($, $p) => ['list', $.map(($) => (0, exports.Group_Part)($, {
        'value serializers': $p['value serializers'],
    }))];
exports.Group = Group;
const Group_Part = ($, $p) => ['state', _pa.cc($, ($) => {
        switch ($[0]) {
            case 'block': return _pa.ss($, ($) => ({
                'state': "block",
                'value': ['text', ({
                        'delimiter': ['quote', null],
                        'value': $,
                    })],
            }));
            case 'nested block': return _pa.ss($, ($) => ({
                'state': "nested block",
                'value': (0, exports.Block)($, {
                    'value serializers': $p['value serializers'],
                }),
            }));
            case 'nothing': return _pa.ss($, ($) => ({
                'state': "nothing",
                'value': ['nothing', null],
            }));
            case 'optional': return _pa.ss($, ($) => ({
                'state': "optional",
                'value': ['optional', $.transform(($) => ['set', (0, exports.Group_Part)($, {
                            'value serializers': $p['value serializers'],
                        })], () => ['not set', null])],
            }));
            case 'sub group': return _pa.ss($, ($) => ({
                'state': "sub group",
                'value': (0, exports.Group)($, {
                    'value serializers': $p['value serializers'],
                }),
            }));
            default: return _pa.au($[0]);
        }
    })];
exports.Group_Part = Group_Part;
const Node = ($, $p) => ['state', _pa.cc($, ($) => {
        switch ($[0]) {
            case 'directory': return _pa.ss($, ($) => ({
                'state': "directory",
                'value': (0, exports.Directory)($, {
                    'value serializers': $p['value serializers'],
                }),
            }));
            case 'file': return _pa.ss($, ($) => ({
                'state': "file",
                'value': (0, exports.Group)($, {
                    'value serializers': $p['value serializers'],
                }),
            }));
            default: return _pa.au($[0]);
        }
    })];
exports.Node = Node;
