// exupery-core-async/dist/algorithms/query/create_unguaranteed_query.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.__create_unguaranteed_query = __create_unguaranteed_query;
const create_guaranteed_query_1 = require("./create_guaranteed_query");
class Unguaranteed_Query_Result_Class {
    constructor(executer) {
        this.executer = executer;
    }
    map_(handle_value) {
        return new Unguaranteed_Query_Result_Class({
            'execute': (on_value, on_exception) => {
                this.executer.execute(($) => {
                    on_value(handle_value($));
                }, on_exception);
            }
        });
    }
    then(handle_value) {
        return new Unguaranteed_Query_Result_Class({
            'execute': (new_on_value, new_on_exception) => {
                this.executer.execute(($) => {
                    handle_value($).__start(new_on_value);
                }, new_on_exception);
            }
        });
    }
    then_unguaranteed(handle_value) {
        return new Unguaranteed_Query_Result_Class({
            'execute': (new_on_value, new_on_exception) => {
                this.executer.execute(($) => {
                    handle_value($).__start(new_on_value, new_on_exception);
                }, new_on_exception);
            }
        });
    }
    map_exception_(handle_exception) {
        return new Unguaranteed_Query_Result_Class({
            'execute': (on_value, on_exception) => {
                this.executer.execute(on_value, ($) => {
                    on_exception(handle_exception($));
                });
            }
        });
    }
    catch_(handle_exception) {
        return (0, create_guaranteed_query_1.__create_guaranteed_query)({
            'execute': (on_value) => {
                this.executer.execute(on_value, ($) => {
                    handle_exception($).__start(on_value);
                });
            }
        });
    }
    __start(on_value, on_exception) {
        this.executer.execute(on_value, on_exception);
    }
}
/**
 * returns an {@link Async_Value }
 * @param executer the function that produces the eventual value
 * @returns
 */
function __create_unguaranteed_query(executer) {
    return new Unguaranteed_Query_Result_Class(executer);
}
